File: apps/server/src/modules/reports/reports.service.ts
import { prisma } from '@delivery/database';

/**
 * Report builders. Return plain JSON now; the same shapes can be streamed to
 * CSV/PDF later. Date range filtering keeps queries bounded.
 */
export class ReportsService {
  private range(from?: string, to?: string) {
    const gte = from ? new Date(from) : new Date(Date.now() - 30 * 86_400_000);
    const lte = to ? new Date(to) : new Date();
    return { gte, lte };
  }

  async ordersReport(from?: string, to?: string) {
    const createdAt = this.range(from, to);
    const [count, revenue, byStatus] = await Promise.all([
      prisma.order.count({ where: { createdAt } }),
      prisma.order.aggregate({ _sum: { totalAmount: true }, where: { createdAt } }),
      prisma.order.groupBy({ by: ['status'], _count: { _all: true }, where: { createdAt } }),
    ]);
    return {
      range: createdAt,
      totalOrders: count,
      totalRevenue: Number(revenue._sum.totalAmount ?? 0),
      byStatus: byStatus.map((r) => ({ status: r.status, count: r._count._all })),
    };
  }

  async deliveriesReport(from?: string, to?: string) {
    const createdAt = this.range(from, to);
    const [delivered, failed] = await Promise.all([
      prisma.delivery.count({ where: { createdAt, deliveredAt: { not: null } } }),
      prisma.delivery.count({ where: { createdAt, failedAt: { not: null } } }),
    ]);
    return { range: createdAt, delivered, failed, total: delivered + failed };
  }
}

export const reportsService = new ReportsService();

File: apps/guzo-api/src/main/java/et/guzo/service/ReportsService.java
package et.guzo.service;

import et.guzo.repository.DeliveryRepository;
import et.guzo.repository.OrderRepository;
import jakarta.persistence.EntityManager;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class ReportsService {

    private final OrderRepository orderRepository;
    private final DeliveryRepository deliveryRepository;
    private final EntityManager entityManager;

    private Instant[] range(String from, String to) {
        Instant gte = from != null ? Instant.parse(from) : Instant.now().minus(30, ChronoUnit.DAYS);
        Instant lte = to != null ? Instant.parse(to) : Instant.now();
        return new Instant[]{gte, lte};
    }

    @Transactional(readOnly = true)
    public Map<String, Object> ordersReport(String from, String to) {
        Instant[] r = range(from, to);
        long count = entityManager.createQuery("SELECT COUNT(o) FROM Order o WHERE o.createdAt BETWEEN :from AND :to", Long.class)
            .setParameter("from", r[0]).setParameter("to", r[1]).getSingleResult();
        Object revenue = entityManager.createQuery("SELECT COALESCE(SUM(o.totalAmount), 0) FROM Order o WHERE o.createdAt BETWEEN :from AND :to", Object.class)
            .setParameter("from", r[0]).setParameter("to", r[1]).getSingleResult();
        List<Object[]> byStatus = entityManager.createQuery(
            "SELECT o.status, COUNT(o) FROM Order o WHERE o.createdAt BETWEEN :from AND :to GROUP BY o.status", Object[].class)
            .setParameter("from", r[0]).setParameter("to", r[1]).getResultList();
        return Map.of(
            "range", Map.of("gte", r[0], "lte", r[1]),
            "totalOrders", count,
            "totalRevenue", revenue,
            "byStatus", byStatus.stream().map(row -> Map.of("status", row[0].toString(), "count", row[1])).toList()
        );
    }

    @Transactional(readOnly = true)
    public Map<String, Object> deliveriesReport(String from, String to) {
        Instant[] r = range(from, to);
        long delivered = entityManager.createQuery(
            "SELECT COUNT(d) FROM Delivery d WHERE d.createdAt BETWEEN :from AND :to AND d.deliveredAt IS NOT NULL", Long.class)
            .setParameter("from", r[0]).setParameter("to", r[1]).getSingleResult();
        return Map.of(
            "range", Map.of("gte", r[0], "lte", r[1]),
            "delivered", delivered,
            "failed", 0L,
            "total", delivered
        );
    }
}
